packages/devtools/lib/src/timeline/frame_flame_chart.dart

```diff
@@ -265,84 +273,51 @@ class FrameFlameChart extends CoreElement {
     return maxRight;
   }
 
-  void _handleMouseMove(MouseEvent e) {
-    // Subtract offset so that [currentMouseX] reflects the x coordinate within
-    // the bounds of the flame chart.
-    _currentMouseX = e.client.x - element.offsetLeft;
-
-    // Store current scroll values for re-calculating scroll location on zoom.
-    _currentScrollLeft = element.scrollLeft;
-    _currentScrollWidth = element.scrollWidth;
-  }
-
   void _handleMouseWheel(WheelEvent e) {
     e.preventDefault();
 
-    if (e.deltaY.abs() > e.deltaX.abs()) {
-      _handleZoom(e.deltaY);
-    } else if (e.deltaX.abs() > e.deltaY.abs()) {
+    if (e.deltaY.abs() >= e.deltaX.abs()) {
+      final mouseX = e.client.x - element.getBoundingClientRect().left;
+      _handleZoom(e.deltaY, mouseX);
+    } else {
       // Manually perform horizontal scrolling.
       element.scrollLeft += e.deltaX;
-      _currentScrollLeft = element.scrollLeft;
-    }
-  }
-
-  void _handleZoom(num deltaY) {
-    // TODO(kenzie): use deltaY to calculate [_zoomMultiplier].
-    if (deltaY > -0.0) {
-      _zoomIn();
-    } else if (deltaY < -0.0) {
-      _zoomOut();
     }
   }
 
-  void _zoomIn() {
+  void _handleZoom(num deltaY, num mouseX) {
     assert(_frame != null);
-    if (_zoomLevel == _maxZoomLevel) {
-      // Already at max zoom level. Do nothing.
-      return;
-    }
 
-    if (_zoomLevel + _zoomMultiplier <= _maxZoomLevel) {
-      _zoomLevel += _zoomMultiplier;
-    } else {
-      _zoomLevel = _maxZoomLevel;
-    }
-    _updateChartForZoom();
-  }
+    deltaY = deltaY.clamp(-maxScrollWheelDelta, maxScrollWheelDelta);
+    num zoomLevel = _zoomLevel + deltaY * _zoomMultiplier;
+    zoomLevel = zoomLevel.clamp(_minZoomLevel, _maxZoomLevel);
 
-  void _zoomOut() {
-    assert(_frame != null);
-    if (_zoomLevel == _minZoomLevel) {
-      // Already at min zoom level. Do nothing.
-      return;
+    if (zoomLevel == _zoomLevel) return;
+    // Store current scroll values for re-calculating scroll location on zoom.
+    double lastScrollLeft = element.scrollLeft.toDouble();
+    // Test whether the scroll offset has changed by more than rounding error
+    // since the last time an exact scroll offset was calculated.
+    if ((floatingPointScrollLeft - lastScrollLeft).abs() < 0.5) {
+      lastScrollLeft = floatingPointScrollLeft;
     }
-
-    if (_zoomLevel - _zoomMultiplier >= _minZoomLevel) {
-      _zoomLevel -= _zoomMultiplier;
+    // Position in the zoomable coordinate space that we want to keep fixed.
+    final double fixedX = mouseX + lastScrollLeft - flameChartInset;
```

Comment:
your code was probably about right once the flameChartInset area that doesn't zoom is factored in.